### ISSUE TITLE
Implications of using EqualityComparer<TValue>.Default in ConcurrentDictionary

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#84163]
Hi! I am studying the [source code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs) of the [`ConcurrentDictionary<TKey, TValue>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2) class, and I noticed the use of the [`EqualityComparer<TValue>.Default`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.equalitycomparer-1.default) inside the `TryUpdateInternal` and `TryRemoveInternal` methods. The `TryUpdateInternal` is called inside the public `AddOrUpdate` method, as part of the optimistic transactional logic around the `updateValueFactory` delegate ([source code](https://github.com/dotnet/runtime/blob/d0d715447aa6957c613109abc6ef874e962c59eb/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1303)). This strikes me as odd. It's not obvious to the consumer of the class that the `TValue` is going to be checked for equality, so they might not pay attention at ensuring that the `TValue` is comparable efficiently. The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate) of the `AddOrUpdate` method does not contain any such hint. Other APIs, like the `TryUpdate`, make it obvious by having a `comparisonValue` parameter, but the `AddOrUpdate` doesn't.

In the extreme case of a `TValue` that is never equal to anything, a type with zero self-awareness, the `AddOrUpdate` halts. To program below enters in an infinite loop and never completes:

```C#
using System;
using System.Collections.Concurrent;

ConcurrentDictionary<int, Insane> dict = new();
dict.TryAdd(1, new());
dict.AddOrUpdate(1, _ => new(), (_, __) => new()); // <=== Halts
Console.WriteLine("Updated OK");

class Insane
{
    public override bool Equals(object obj) => false;
}
```

[Online demo](https://dotnetfiddle.net/ZGcliE).

I am showing the above demo as a curiosity. No one would implement the `Equals` like this, and if they did the consequences would be immediately obvious. Such code would never make it to production. But it's not difficult to slip into production code with a value-type `TValue` that doesn't override the `Equals`, resulting in [reduced performance and increased allocations](https://devblogs.microsoft.com/premier-developer/performance-implications-of-default-struct-equality-in-c/ "Performance implications of default struct equality in C#"). Or with a reference-type `TValue` with a custom and heavy `Equals` implementation (think of `SequenceEqual` or other looping code).

I would like to ask if it would be a good idea to change the logic of the `AddOrUpdate` so that it compares `Node` objects instead the `TValue`s, in cases that doing so is safe. My understanding is that it should be safe for `TValue`s that fall into the category `typeof(TValue).IsValueType && !ConcurrentDictionaryTypeProps<TValue>.IsWriteAtomic`, because in this case the nodes are replaced every time the `TValue` is modified. As for value types that can be written atomically, these belong to a predefined group of types with well known equality behavior ([source code](https://github.com/dotnet/runtime/blob/d0d715447aa6957c613109abc6ef874e962c59eb/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L2313)), so using the `EqualityComparer<TValue>.Default` should be OK. And finally for reference types, I think that comparing the `TValue`s with the straightforward `ReferenceEquals` should be safer than invoking whatever unknown code can reside in the `Equals` of those types.

Thanks!

**Comment by eiriktsarpalis** [comment]
It's an interesting point, I notice for example that the `TryRemoveInternal` method does expose a [`matchValue` flag](https://github.com/dotnet/runtime/blob/d0d715447aa6957c613109abc6ef874e962c59eb/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L388) that only enables this behaviour [where appropriate](https://github.com/dotnet/runtime/blob/d0d715447aa6957c613109abc6ef874e962c59eb/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L378).

I suspect it could be possible to remove the dependency on `EqualityComparer<TValue>.Default` in the `TryUpdateInternal` method (either by hardcoding reference equality on `TValue` directly for reference types or using some kind of versioning scheme in the containing `Node` class), but at this point I think that it would be a potential breaking change for scenaria where both `TValue` has nonstandard equality and the update delegate uses side-effects.

> In the extreme case of a TValue that is never equal to anything, a type with zero self-awareness, the AddOrUpdate halts.

Users should expect undefined behavior when providing implementations of equality that don't satisfy the [basic axioms for equality](https://en.wikipedia.org/wiki/Equivalence_relation). Similar to how a binary search trees stop working when you specify an `IComparer<T>` that isn't a total order.

cc @stephentoub

**Comment by stephentoub** [comment]
This would be a breaking change, I believe the current implementation is doing the right thing even if it could be misused, and in over a decade with CD I've not heard of anyone having a problem because of this. If there's opportunity to improve performance without any behavioral impact, great, but I'm not seeing that here.

**Comment by theodorzoulias** [comment]
> but at this point I think that it would be a potential breaking change for scenaria where both `TValue` has nonstandard equality and the update delegate uses side-effects.

@eiriktsarpalis do you mean that some callers might expect that the `TValue.Equals` will be called as a side-effect of calling the `AddOrUpdate`, and their existing code will break in case the `TValue.Equals` is not called?

**Comment by eiriktsarpalis** [comment]
Yes, either intentionally or accidentally.

**Comment by theodorzoulias** [comment]
> I suspect it could be possible to remove the dependency on `EqualityComparer<TValue>.Default` in the `TryUpdateInternal` method (either by hardcoding reference equality on `TValue` directly for reference types or using some kind of versioning scheme in the containing `Node` class)

I thought about the versioning idea, but I can't make it work safely in my mind. For example let's say that an `int _version` field is added in the [`Node`](https://github.com/dotnet/runtime/blob/958b1e046cbfff5c0bc647f531b4fcf9eb197a6c/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L2203) class, and the `AddOrUpdate` logic relies on this field to know if the value has changed. It is now possible that the `_version` will wrap, resulting is erroneous behavior. For example one thread invokes the `updateValueFactory`, which takes a long time to complete, and in the meantime other threads update the same key exactly 4,294,967,296 times. Then the initial thread will overwrite the changed `_value` field, because the `_version` field will be exactly the same as before...

**Comment by stephentoub** [comment]
> the TValue.Equals will be called as a side-effect

It's not just about side effects from invoking the method.  Two types can compare equally even if they aren't the same object or the same bit pattern, and whether Equals is true or not dictates whether the value will be replaced in the dictionary, so changing this to something like reference equality has a material impact on what objects are exposed from the dictionary.

> The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate) of the AddOrUpdate method does not contain any such hint.

The documentation for TryUpdate does:
"Updates the value associated with key to newValue if the existing value with key is equal to comparisonValue."
"The value that replaces the value of the element that has the specified key if the comparison results in equality."
"The value that is compared with the value of the element that has the specified key."
"true if the value with key was equal to comparisonValue and was replaced with newValue; otherwise, false."

The docs for AddOrUpdate could be clarified similarly.

**Comment by theodorzoulias** [comment]
> It's not just about side effects from invoking the method. Two types can compare equally even if they aren't the same object or the same bit pattern, and whether Equals is true or not dictates whether the value will be replaced in the dictionary, so changing this to something like reference equality has a material impact on what objects are exposed from the dictionary.

@stephentoub what's the worst that could happen? The worst scenario that I can think of is this:

- The `TValue` is a reference type.
- The current thread gets the `oldValue` and invokes the `updateValueFactory` delegate.
- In the meantime other threads change the `_value` back and forth, so that it becomes a different object that the `oldValue`, but equal according to the `TValue.Equals` implementation.
- The current thread completes the `updateValueFactory`, and compares the `oldValue` with the stored value.
- **Existing behavior:** The current thread replaces the stored value with the value produced by the `updateValueFactory` (because the `Equals` succeeds).
- **Proposed behavior:** The current thread invokes the `updateValueFactory` again, with the stored value as argument (because the `ReferenceEquals` fails).

So the worst scenario is that occasionally the `AddOrUpdate` will invoke the `updateValueFactory` one extra time or two. In either case the stored value will be replaced. The value returned by the `AddOrUpdate` will always be the result of the last `updateValueFactory` invocation.

**Comment by theodorzoulias** [comment]
> This would be a breaking change, I believe the current implementation is doing the right thing even if it could be misused, and in over a decade with CD I've not heard of anyone having a problem because of this.

Personal story: recently I changed an `AsyncKeyedLock<T>` [implementation](https://stackoverflow.com/questions/31138179/asynchronous-locking-based-on-a-key/65256155#65256155), from being based on a 
`Dictionary<TKey, (SemaphoreSlim, int)>` to a `ConcurrentDictionary<TKey, Entry>`. The `Entry` being a `readonly struct` with a couple of fields (a `SemaphoreSlim` and an `int`). Just for funsies I declared the struct as `record`, for no particular reason other than for playing with the latest C# 10.0 goodies. Actually I expected that a `record` would be a bit slower, because the members of records are properties instead of fields. Little did I know that this change improved significantly the performance of my class, because `record`s implement the `IEquatable<T>` interface while my initial naked `readonly struct` implementation did not. I wouldn't be surprised if some people are paying unknowingly for years the price of reflection in their apps, because they can't imagine that the type of the `TValue` impacts the performance of the `ConcurrentDictionary<K,V>` collection. Personally I am familiar with this class for a long time, I have used it many times the last 3-4 years, and this realization come up to me just yesterday.
